src/Signal/Pcntl.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+namespace React\EventLoop\Signal;
+
+use React\EventLoop\LoopInterface;
+use React\EventLoop\Timer\TimerInterface;
+
+/**
+ * @internal
+ */
+final class Pcntl
+{
+    /**
+     * @var LoopInterface
+     */
+    private $loop;
+
+    /**
+     * @var array
+     */
+    private $signals = [];
+
+    /**
+     * @var TimerInterface|null
+     */
+    private $timer;
+
+    /**
+     * @var bool
+     */
+    private $enabled = false;
+
+    /**
+     * @param LoopInterface $loop
+     */
+    public function __construct(LoopInterface $loop)
+    {
+        $this->loop = $loop;
+        $this->enabled = \extension_loaded('pcntl');
+    }
+
+    /**
+     * Register the given $listener with the
+     * given $signal. When this is the first
+     * listener for $signal add a listener
+     * for $signal, when this isn't the
+     * first add it to the list of registered
+     * listeners. When this is the first
+     * listener for all signals also install
+     * a timer to check for incoming signals.
+     *
+     * @param int $signal
+     * @param callable $listener
+     *
+     * @return void
+     */
+    public function on($signal, callable $listener)
+    {
+        if (!$this->enabled) {
+            return;
+        }
+
+        if ($this->timer === null) {
+            $this->timer = $this->loop->addPeriodicTimer(0.1, function () {
```

Comment:
Calling `pcntl_signal_dispatch()` can simply be rolled into the main event loop after calling `stream_select()`, since the latter is interrupted by the arrival of signals, regardless of the timeout parameter.